### ISSUE TITLE
Nrgnn: Learning a label noise resistant graph neural network on sparsely and noisily labeled graphs

### ISSUE THREAD

**Issue by 2nazero** [AIML-K/GNN_Survey#17]

**Comment by 2nazero** [comment]
https://github.com/EnyanDai/NRGNN

**Comment by 2nazero** [comment]
<img width="324" alt="스크린샷 2024-11-02 오후 5 00 36" src="https://github.com/user-attachments/assets/6ab3df79-4631-4b6a-bddf-f1e70fcc19d8">

This is the main figure that explains about this paper.

**Comment by 2nazero** [comment]
## Methodology

### Edge Predictor
The edge predictor identifies and adds missing links between nodes in the graph.
- **Node Representation**: Each node representation $Z$ is learned using a GCN based on the initial adjacency matrix $A$ and node features $X$:
  
$$Z = \text{GCN}(A, X)$$

where $Z$ contains embeddings that capture both **feature and structural information**.

- **Link Prediction**: For two nodes $i$ and $j$, the probability $S_{ij}$ that they are linked is calculated as:

 $$S_{ij} = \sigma(z_i^\top z_j)$$
 
where $z_i$ and $z_j$ are the embeddings of nodes $i$ and $j$, and $\sigma$ is an activation function, such as ReLU, to prevent gradient vanishing issues.

- **Loss Function**: The edge predictor minimizes the **adjacency matrix reconstruction loss**, where for each positive sample $S_{ij} = 1$, a negative sample (unlinked nodes) is randomly selected. The objective is to ensure the edge predictor can accurately predict missing edges:

$$\min_{Q_N} L_N = \sum_{i,j \in \text{linked}} (S_{ij} - 1)^2 + \sum_{i,j \notin \text{linked}} (S_{ij})^2$$

### Pseudo Label Miner
The pseudo label miner assigns pseudo labels to unlabeled nodes, thereby expanding the effective label set and providing additional learning signals.

- **Link-based Pseudo Label Assignment**: The pseudo label miner uses the updated graph with predicted links, where nodes $i$ and $j$ are connected if the similarity score $S_{ij}$ exceeds a threshold $L$:

$$S_{ij} = \begin{cases} 
      1, & \text{if } S_{ij} > L \\
      0, & \text{otherwise} 
   \end{cases}$$

- **Pseudo Label Prediction**: The pseudo label miner uses a GNN classifier trained on the newly structured graph. For each unlabeled node $i$, the pseudo label $\hat{y}_i$ is:

$$\hat{y}_i = \text{GNN}(S_O, X)$$

  where $S_O$ is the adjacency matrix with the newly added edges, and $X$ is the feature matrix.

- **Pseudo Label Selection**: High-confidence pseudo labels are selected based on a threshold. The set of predictions from the pseudo label miner is filtered to ensure only reliable labels are added to the extended label set.

### GNN Classifier
The GNN classifier is trained on the updated graph with both the newly created edges and the expanded label set. 

- **Objective Function**: The loss function for the GNN classifier incorporates supervision from both the **noisy labeled data** and the **pseudo-labeled nodes**, ensuring that the classifier is optimized to handle label noise:

$$L_G = \sum_{i \in V_M} \text{CrossEntropy}(y_i, \hat{y}_i)$$

  where $V_M$ is the set of nodes with either noisy or pseudo labels, $y_i$ is the label (noisy or pseudo), and $\hat{y}_i$ is the prediction.

### Overall Training Objective
The **joint objective** function for optimizing the edge predictor $Q_N$, pseudo label miner $Q_M$, and GNN classifier $Q_G$ is given by:

$$\min_{Q_N, Q_M, Q_G} L_G + \alpha L_N + \beta L_M$$

where $L_N$ is the edge prediction loss, $L_M$ is the pseudo label mining loss, $\alpha$ and $\beta$ are hyperparameters controlling the balance between these terms, and $L_G$ is the node classification loss.

**Comment by 2nazero** [comment]
## Experiments

Main Questions to Answer
1. Is NRGNN robust to different types and levels of label noise?
2. How does NRGNN perform with varying sizes of noisy labels and different levels of graph sparsity?
3. How flexible is NRGNN with different GNN backbones, and how do the edge predictor and pseudo label miner contribute to its performance?

**Comment by 2nazero** [comment]
root@e69f0e0aedb4:~/NRGNN# bash train_NRGNN.sh
/opt/conda/lib/python3.8/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)
  return torch._C._cuda_getDeviceCount() > 0
cpu
Namespace(alpha=0.03, beta=1, cuda=False, dataset='cora', debug=False, dropout=0.5, edge_hidden=64, epochs=200, hidden=16, label_rate=0.05, lr=0.001, n_n=50, n_p=-1, no_cuda=False, noise='uniform', p_u=0.8, ptb_rate=0.2, seed=10, t_small=0.1, weight_decay=0.0005)
Loading cora dataset...
Selecting 1 largest connected components
Uniform noise
Actual noise 0.16
Optimization Finished!
Total time elapsed: 193.8855s
picking the best model according to validation performance
=====validation set accuracy=======
        Predictor results: loss= 1.2199 accuracy= 0.8273
        GCN classifier results: loss= 1.0909 accuracy= 0.8434
===================================
=====test set accuracy=======
        Predictor results: loss= 1.2596 accuracy= 0.7857
        GCN classifier results: loss= 1.1463 accuracy= 0.8003
===================================
/opt/conda/lib/python3.8/site-packages/torch/cuda/__init__.py:52: UserWarning: CUDA initialization: Found no NVIDIA driver on your system. Please check that you have an NVIDIA GPU and installed a driver from http://www.nvidia.com/Download/index.aspx (Triggered internally at  /pytorch/c10/cuda/CUDAFunctions.cpp:100.)
  return torch._C._cuda_getDeviceCount() > 0
cpu
Namespace(alpha=0.3, beta=1, cuda=False, dataset='cora', debug=False, dropout=0.5, edge_hidden=64, epochs=200, hidden=16, label_rate=0.05, lr=0.001, n_n=50, n_p=-1, no_cuda=False, noise='pair', p_u=0.8, ptb_rate=0.2, seed=10, t_small=0.1, weight_decay=0.0005)
Loading cora dataset...
Selecting 1 largest connected components
Pair noise
Actual noise 0.16

**Comment by 2nazero** [comment]
# Code Comparison 2 - Results w/o Codes
<img width="439" alt="스크린샷 2024-11-07 오후 4 58 37" src="https://github.com/user-attachments/assets/e42397c3-46f9-4970-a3b2-0fc13eee3f6b">
Results of ablation study needs modifications of the codes.
